src/pad/projectAutoDetectionManager.ts

```diff
@@ -0,0 +1,157 @@
+          '**/node_modules/**'
+        );
+    return promise.then(
+      results => this.detect(results, projects),
+      error => ErrorHandler.logError(error)
+    );
+  }
+
+  private detect(
+    results: models.IVSCodeUri[],
+    projects: models.Projects[]
+  ): models.IProjectDetectionResult {
+    if (!results || !results.length) {
+      return;
+    }
+    let detectionResult: models.IProjectDetectionResult;
+    for (const project of projects) {
+      detectionResult = this.checkForProject(results, project);
+      if (detectionResult.apply) {
+        break;
+      }
+    }
+    return detectionResult;
+  }
+
+  private checkForProject(
+    results: models.IVSCodeUri[],
+    project: models.Projects
+  ): models.IProjectDetectionResult {
+    let presetName: string;
+    switch (project) {
+      case models.Projects.angular:
+        presetName = constants.vsicons.presets.angular;
+      default:
+        break;
+    }
+
+    // NOTE: User setting (preset) bypasses detection in the following cases:
+    // 1. Preset is set to 'false' and icons are not present in the manifest file
+    // 2. Preset is set to 'true' and icons are present in the manifest file
+    // For this cases PAD will not display a message
+    // We need to check only the 'workspaceValue' ('user' setting should be ignored)
+    const iconsDisabled: boolean = ManifestReader.iconsDisabled(project);
+    const preset = this.configManager.getPreset<boolean>(presetName)
+      .workspaceValue;
+    const bypass =
+      preset != null &&
+      ((!preset && iconsDisabled) || (preset && !iconsDisabled));
+
+    if (bypass) {
+      return { apply: false };
+    }
+
+    // We need to mandatory check the following:
+    // 1. The project related icons are present in the manifest file
+    // 2. It's a detectable project
+    // 3. The preset (when it's defined)
+
+    // Use case: User has the preset set but project detection does not detect that project
+    const projectInfo: models.IProjectInfo = this.getProjectInfo(
+      results,
+      project
+    );
+    const enableIcons = iconsDisabled && (!!projectInfo || preset === true);
+    const disableIcons = !iconsDisabled && (!projectInfo || preset === false);
+
+    if (!enableIcons && !disableIcons) {
+      return { apply: false };
+    }
+
+    const langResourceKey: models.LangResourceKeys = enableIcons
+      ? projectInfo
+        ? models.LangResourceKeys.ngDetected
+        : models.LangResourceKeys.nonNgDetectedPresetTrue
+      : projectInfo
+        ? models.LangResourceKeys.nonNgDetected
+        : models.LangResourceKeys.ngDetectedPresetFalse;
+
+    return {
+      apply: true,
+      projectName: project,
+      langResourceKey,
+      value: enableIcons || !disableIcons,
+    };
+  }
+
+  private getProjectInfo(
+    results: models.IVSCodeUri[],
+    project: models.Projects
+  ): models.IProjectInfo {
+    let projectInfo: models.IProjectInfo = null;
+    results.some(result => {
```

Comment:
Strange usage of the `some` method by introducing side-effects. I would refactor this by using `find` instead. You get the same short-circuit behavior and no side-effects inside the closure.

Reply:
It has the meaning of `if any`. Will also check behavior with `find`.

Reply:
Is this more appropriate?
```ts
    let projectInfo: models.IProjectInfo = null;
    results.find(result => {
      const content: string = readFileSync(result.fsPath, 'utf8');
      const projectJson: { [key: string]: any } = Utils.parseJSON(content);
      projectInfo = this.getInfo(projectJson, project);
      return !!projectInfo;
    });
```

Reply:
I was more thinking of both using `map` and `find` but this would make the short-circuit to be only applied in the second iteration:

```ts
const projectInfo = results.map(result => {
      const content: string = readFileSync(result.fsPath, 'utf8');
      const projectJson: { [key: string]: any } = Utils.parseJSON(content);
      return this.getInfo(projectJson, project);
}).find(x => !!projectInfo);
```

In case performance is relevant I would prefer a normal `for loop`. I don't like the idea of introducing side-effects in `map-like` kind of methods.

Reply:
I'm in favor of the `for loop` too.